### ISSUE TITLE
Metadata Block for MIME Type

### ISSUE THREAD

**Issue by sjehuda** [Tampermonkey/tampermonkey#1498]
Hello,
Apparently, my request [was not accepted earlier](https://github.com/greasemonkey/greasemonkey/issues/3137).
I hope you would find it reasonable as I do.

There are userscripts that intended for [XML documents](https://github.com/greasemonkey/greasemonkey/issues/3135) or HTML documents, or even [Image documents](https://github.com/Owyn/HandyImage) , some exclusively and some for multiple types.

Currently,  I use `Document.contentType` (e.g. `document.contentType.match('xml')`) to determine whether or not script should start.

However, a better way to find out the filetype, is by `XMLHttpRequest`.
```
xmlHTTPRequest = new XMLHttpRequest();
xmlHTTPRequest.open('GET', document.documentURI, false);
xmlHTTPRequest.send();
xmlHTTPRequest.getResponseHeader('Content-Type');
```
But, currently, it's not allowed (deprecated) to make such request on the main thread unless it's called via callback, which is fairly complexed, or inside a so-called Web Worker, which must be inside a separate JS file.

Surely, we can call a file as string and manipulate it, but it's not a proper solution, IMHO.

A built-in functionality (extension level) that is able to store filetype upon request is vital, of course only when that metadata block is used.

Metadata block API can make this task easier for scriptwriters and make a friendlier user experience.

In my case, it will also help to be avoided from being blocked by server due to "surge of (http) requests". And considering the nature of Greasemonkey, this feature is a must! ([read example](https://github.com/greasemonkey/greasemonkey/issues/3137#issuecomment-1108935352))

**Comment by 7nik** [comment]
>  intended for XML documents 

links to his own script. On user script hosting sites I found around one hundred scripts that include the word XML but most, if not almost all, of them (I'm too lazy to read the description of each of them) just export some data in XML format or something and still intended to run on HTML pages. Thus, there is a very small amount of scripts designed to run on XML pages.

> Image documents 

links to a userscript that runs on ordinary HTML pages of image hosters. It even does early return when is run on an image page.

> *makes synchronous XHR*

man, don't do such terrible things. I hope you understand why it isn't allowed in the main thread.
In TM you can use top-level await. Not sure about other script managers, but anyway you can wrap the code with async IIFE.

```js
const contentType = await fetch(document.documentURI).then(resp => resp.headers.get("Content-Type"));
if (!contentType.includes("xml")) return;
```
>  being blocked by server due to "surge of (http) requests"

no one will block for one extra request. The very lowest border is something like 5 req/sec but often it is dozens of requests per second and during a noticeable period of time.

> When we run a check using "document" we might get
document.contentType
"text/html"

Have you read the docs of `document.contenType`? From docs:
> the MIME type that the document is being rendered as. This may come from HTTP headers or other sources of MIME information, and might be affected by automatic type conversions performed by either the browser or extensions.

**Comment by sjehuda** [comment]
Thank you for providing code.  I didn't try it, though I might test it very soon.
FYI, we've taken care of it and the [script](https://bugs.kde.org/show_bug.cgi?id=449735) now works without warnings.

Concerning this issue:
In my [second post](https://github.com/greasemonkey/greasemonkey/issues/3137#issuecomment-1108935352), where my suggestion was rejected, I gave an example which shows that `document.contenType` and `xmlHTTPRequest.getResponseHeader('Content-Type')` give different results.

**Comment by sjehuda** [comment]
@7nik please take this feed: https://gist.github.com/7nik.atom

- `xmlHTTPRequest.getResponseHeader('Content-Type')` outputs `application/atom+xml; charset=utf-8`
- `document.contentType` outputs `"text/plain"`

The size of these gist feeds (text alone) can reach even to 5MB and that will result in downloading 10MB .

I'm against that.

Is there a way to use xmlHTTPRequest without downloading the entire file (i.e. just header)?
If not, then I think this feature is vital.

**Comment by sjehuda** [comment]
Please have a look at the userscript https://openuserjs.org/scripts/sjehuda/Newspaper

**Comment by 7nik** [comment]
I don't say that the header Content-Type and `document.contentType` are the same, even more, I quoted the docs why they are different.

You can try to use the [HEAD method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/HEAD), but you need to test how reliable it is or use the GET method as a fallback.

**Comment by sjehuda** [comment]
I'm sorry.Thank you, HEAD parameter seems to be what I was looking for.

I didn't apply that you did.
I wanted to elaborate on that.

Thank you, HEAD parameter seems to be what I was looking for.

If I correct, I need to change `@run-at` to `document-start` and call twice to XMLHTTPRequest, one with HEAD and if page is a Feed, then once again with GET.

I'll look into it and report.
Thank you

**Comment by sjehuda** [comment]
Thank you @7nik for your help.

**Comment by 7nik** [comment]
Just use `fetch`, it is more simple.
This should work:
```js
async function httpGetAsync(uri, callback) {
  let resp = await fetch(uri, { method: "HEAD" });
  if (!resp.ok) resp = await fetch(uri);
  const mimeType = resp.headers.get("Content-Type");
  callback(mimeType);
}
```

**Comment by sjehuda** [comment]
Thank you @7nik

I've studied fetch, just today.
It's nice to have.

I've wrote an observation at [Newspaper/issues/Observation_Fetch_API](https://openuserjs.org/scripts/sjehuda/Newspaper/issues/Observation_Fetch_API).

Because fetch doesn't support `file://` URL, I've decided not to use it.
I do consider using `Promise` with `XMLHttpRequest`.

**Comment by sjehuda** [comment]
This is the Promise API code.
It works for HTTP links.
It doesn't seem to reach to the statement `if (document.URL.startsWith('file:'))`.
Unless there is a way, I'll stay with `callback` and `XMLHttpRequest`.

```javascript
let myPromise = new Promise(function(myResolve, myReject) {
  let request = new XMLHttpRequest();
  request.open('GET', document.documentURI);
  request.onload = function() {
    if (request.status == 200) {
      myResolve(request);
    } else {
      myReject("File not Found");
    }
  };
  request.send();
});

myPromise.then(
  function(request) {
      if (document.URL.startsWith('file:')) {
        mimeType = document.contentType;
      } else {
          mimeType =
            request
            .getResponseHeader('Content-Type');
        }
      if (checkMediaType(mimeType)) {
        renderFeed(request.response)
      }
  },
  function(error) {checkMediaType(error);}
);
```
